packages/jest-runtime/src/index.ts

```diff
@@ -321,13 +321,10 @@ class Runtime {
     let moduleRegistry;
 
     if (!options || !options.isInternalModule) {
-      if (
-        this._moduleRegistry.get(modulePath) ||
-        !this._isolatedModuleRegistry
-      ) {
-        moduleRegistry = this._moduleRegistry;
-      } else {
+      if (this._isolatedModuleRegistry) {
```

Comment:
For the most part it looks good but there is one thing that worries me.

With this new implementation there is one change in behavior.

Let's say that I have module A which depends in module B. If I do . `jest.isolateModules(() => A = require('./A');)` it will now also use a different copy of "B" also_(as well as for the dependencies of B)_ I'm worried that this might not be the what the user expects, and might make it a bit hard to debug issues.

My example of A and B might be easier to understand with A being `MyComponent` and B being `React`.

thoughts? @SimenB @thymikee

Reply:
Good point. I've updated the code to not affect dependencies within an isolated module

Reply:
Hi!

Thanks for all the work. Sorry, what I meant with my comment was just to express that there would be a change in behavior, I’m not saying that one is right and the other isn’t.


@thymikee @SimenB @willym1 I think it is important define exactly how we want `isolateModules` to work before proceeding.

I see a couple of options:
1. [Current behavior] When an  `isolateModules` block gets executed, only modules that are not in the module registry yet will be isolated. This means that things that have already been required will use the existing “require”.
2. When an  `isolateModules` block gets executed, it isolates every single module that gets required.
3. When an  `isolateModules` block gets executed, only the required only the direct requires get isolated but not their internal dependencies.

Here are thoughts for the each option:
* Rolling out option 2 and 3 will break a lot of existing tests when people upgrade their versions of Jest in ways that might be hard to spot.
* Option 1 still has the problem mentioned in #7863, but for some reason I still I feel this might be the safest behavior for the user(as well as for maintainability). I wonder if we could log a warning when we detect that the user is trying to require a module that is already in the registry. I’m not sure if we can correctly identify those cases without much false positives.

What are your thoughts? @thymikee @SimenB @willym1

Reply:
> This means that things that have already been required will use the existing “require”.

This is not really documented, isn't it?

I think I lean towards 2, since I'd expect everything inside the callback to get its own copy. This is fairly fresh addition to Jest, so I think there won't be too much cruft around it, especially if we're fixing it for good reasons.

Reply:
> This is not really documented, isn't it?

Correct, it is not documented.

Reply:
Leaning towards 2 as well. 1 doesn't seem functionally intuitive. If you needed to do multiple tests on a module that requires each one to be a fresh copy, you'd have to individually isolate each module at the start of the test and know beforehand how many copies are needed.

1 and 3 suffers from this issue - since `isolateModules` calls aren't hoisted like a regular `mock` call would (correct me on this if I'm wrong), there's no guarantee that the internal dependencies of those copies were already previously required. Say you were isolating a module multiple times whose behavior is tied to some stateful internal dependencies. Any copy of the isolated module can indirectly affect the other copies simply by running, which may be problematic if you're trying to do multiple isolated tests. This is just a very specific case btw, I don't see this significantly impacting anyone currently using `isolatedModules` in their tests.

Reply:
@rogeliog In my opinion, option 2 is the most intuitive, and the only behaviour that matches the name of the function. This is exactly what I thought `isolatedModules` did when I read the documentation, especially after having just read the documentation for `resetModules()` which is right above it, and the explicit mention of the word "sandbox".

> This is just a very specific case btw, I don't see this significantly impacting anyone currently using isolatedModules in their tests.

@willym1 I found this issue+PR because this is something I'm currently struggling with. Because of this bug (and because of the lack of async import support: #10428) I've had to use `resetModules()` instead.